File: dsa/cpp/treeLevelOrderleetcode.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Binary tree node definition
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = nullptr;
        right = nullptr;
    }
};

// Function that returns level order traversal as vector of vectors
vector<vector<int>> levelOrderTraversal(Node* root) {
    vector<vector<int>> result;

    if (root == nullptr) return result;

    queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
        int levelSize = q.size();
        vector<int> currentLevel;

        for (int i = 0; i < levelSize; ++i) {
            Node* curr = q.front();
            q.pop();

            currentLevel.push_back(curr->data);

            if (curr->left != nullptr)
                q.push(curr->left);
            if (curr->right != nullptr)
                q.push(curr->right);
        }

        result.push_back(currentLevel);
    }

    return result;
}

int main() {
    // Creating a sample binary tree:
    //
    //        1
    //       / \
    //      2   3
    //     / \   \
    //    4   5   6

    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->right = new Node(6);

    vector<vector<int>> traversalResult = levelOrderTraversal(root);

    // Printing the result
    cout << "Level Order Traversal:\n";
    for (const auto& level : traversalResult) {
        for (int val : level) {
            cout << val << " ";
        }
        cout << endl;
    }

    return 0;
}

File: dsa/python/treeLevelOrderleetcode.py
# Definition for a binary tree node
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Solution class with level order traversal
class Solution(object):
    def levelOrder(self, root):
        if not root:
            return []

        result = []
        queue = [root]

        while queue:
            level = []
            next_queue = []

            for node in queue:
                level.append(node.val)
                if node.left:
                    next_queue.append(node.left)
                if node.right:
                    next_queue.append(node.right)

            result.append(level)
            queue = next_queue

        return result

# ---------- Tree Construction ----------

# Example tree:
#       1
#      / \
#     2   3
#        / \
#       4   5

# Creating nodes
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.right.left = TreeNode(4)
root.right.right = TreeNode(5)

# ---------- Call Level Order ----------

sol = Solution()
output = sol.levelOrder(root)
print("Level Order Traversal:", output)  # Output: [[1], [2, 3], [4, 5]]
